design/observerability-2.md

```diff
@@ -0,0 +1,196 @@
+
+The Quarkus-based distribution (Keycloak.x) requires adding `smallrye-metrics` extension (see [the manual](https://quarkus.io/guides/microprofile-metrics)).
+Once this is done, the metrics will be automatically available in `/metrics` endpoint. From the implementation perspective,
+we need to perform the same steps as we were on Wildfly.
+
+## Metrics for the REST interface
+
+Most of the metrics are related to the various REST endpoints. Therefore, it makes sense to put the root metrics
+component into `services` module under its own package - `org.keycloak.services.metrics`.
+
+During initialization, the root metrics components should obtain an Application `MetricRegistry` from `MetricRegistries`.
+The next step is to register all required `Metric`s using the `#register` method.
+
+The best metric for measuring REST latency is Histogram. It gives the most information about the traffic going through
+the specific REST method.
+
+A bit more interesting is measuring the number of response codes in certain time (e.g. the number of 200 OK within a minute).
+This requires registering at least 4 different counters in Microprofile Metrics: the number of 200s, 300s, 400s and 500s.
+
+The easiest way to measure both latency and response codes is to register a new `MetricsFilter` that implements
+`ContainerRequestFilter` and `ContainerResponseFilter`. The implementation also requires implementing a `MetricsWebFilter`
+that implements a `Filter`. As Smallrye [Javadocs mentions](https://github.com/smallrye/smallrye-metrics/blob/b517d9904f91fd0077e0a215e5bb87ef9f1ac9ee/implementation/src/main/java/io/smallrye/metrics/jaxrs/JaxRsMetricsFilter.java#L36-L43):
+
+> The reason for having two filters (one based on Servlet API, one on JAX-RS API)
+> is to overcome the limitation of JAX-RS specification in that it does not require
+> implementations to call a ContainerResponseFilter to finish exchanges that led
+> to an unmapped exception. That's why we can't use a ContainerResponseFilter to
+> 'finish' the tracking of a REST call and update the corresponding metrics.
+
+it is necessary to properly track all non-successful responses.
+
+The last part is tracking external IdP responses. Implementing this part requires changing the existing code around
+`SocialIdentityProvider`s and `OIDCIdentityProvider`. All `SocialIdentityProvider` implementations usually throw
+`IdentityBrokerException`s. The exception needs to be enhanced to contain an information about that provider has thrown
+it. Then, the implementation should involve updating all places in `AbstractOAuth2IdentityProvider` and add
+proper checks in all `catch` clauses. Once an `IdentityBrokerException` is caught, and it contains a provider ID,
+it should result in updating certain metrics in the metrics components after logging an error.
+
+## Logging architecture
+
+Collecting logs in an OpenShift cluster is a cross-cutting concern that needs a unified solution. OpenShift provides a
```

Comment:
Is this just a log4j logger, or something else?

Reply:
Keycloak uses JBoss Logging as a logging facade. So this code needs to be applied on top of JBoss Logging.